README.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      README.md
--
--  BRIEF
--      Important information regarding this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--
------------------------------------------------------------------------------->

# italiano

## Summary

A very simple translator from Italian to some other languages.

## License

This project's license is **GPL 2** (as of June 1991).  The whole license text
can be found in `LICENSE` in the main directory of this repository.  The brief
version is as follows:

> Copyright (C) 2022 Kevin Matthes
>
> This program is free software; you can redistribute it and/or modify
> it under the terms of the GNU General Public License as published by
> the Free Software Foundation; either version 2 of the License, or
> (at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
>
> You should have received a copy of the GNU General Public License along
> with this program; if not, write to the Free Software Foundation, Inc.,
> 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

When compiling a printable version of this documentation using Pandoc, the full
license will be attached automatically to the resulting document.  This can be
invoked by calling `repository-manual.m`.

## Software Requirements

| Requirement       | Type          | Role                                  |
|:------------------|:-------------:|:--------------------------------------|
| `docs-snippets`   | repository    | documentation constants               |
| Flex              | application   | Lex compiler                          |
| GCC               | application   | C compiler                            |
| GNU Octave        | application   | execution of the provided scripts     |
| Pandoc            | application   | compilation of repository manual      |
| `texlive-full`    | package       | compilation of repository manual      |

The compilation of such an **optional** repository manual can be invoked by just
calling one of the following lines in a terminal.

```
octave repository-manual.m
octave-cli repository-manual.m
```

Both will redirect to the same application, GNU Octave, which will then create
the manual for this repository and attach the entire license to it.  The
resulting file, `repository.pdf`, will be saved in the main directory of this
repository.

## Description

This repository provides a simple lexer in order to translate Italian to a
selected language.  It can be compiled by invoking the GNU Octave script
`flex-gcc.m` in a terminal:

```
octave flex-gcc.m
octave-cli flex-gcc.m
```

The script `install-app.m` will install the application in `~/.local/bin/`, at
option.

```
octave install-app.m
octave-cli install-app.m
```

## Options

### `--american`

```
italiano --american
```

Set the output language to US American English.  Only one target language is
allowed.

### `--british`

```
italiano --british
```

Set the output language to British English.  Only one target language is
allowed.

### `--german`

```
italiano --german
```

Set the output language to German.  Only one target language is allowed.

### `-V`

```
italiano -V
```

Show verbose information.  Can be combined with any option.

### `-i`

```
italiano -i
```

Show a brief information block and quit.  All further options will be ignored.

### `-l`

```
italiano -l
```

Show the license text and quit.  All further options will be ignored.

### `-s`

```
italiano -s
```

Suppress verbose information.  Can be combined with any option.

### `-v`, `--version`

```
italiano -v
italiano --version
```

Show the version and quit.  All further options will be ignored.

### `-w`

```
italiano -w
```

Show the warranty information and quit.  All further options will be ignored.

## Hints

### Unique Word Protection

In order to protect a unique word from being replaced, `%` needs to be given as
a prefix to the respective word.

Due to the defined replacement rules, unknown words might be misunderstood and,
thus, translated incorrectly.  This affects words such as names which contain
a configurated vocab.

```
$ echo %Io | italiano --british
Io
$ echo Io | italiano --british
I
```

<!----------------------------------------------------------------------------->


CHANGELOG.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      CHANGELOG.md
--
--  BRIEF
--      The development history of this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--      See `README.md' for project details.
--
------------------------------------------------------------------------------->

# Changelog

## [Unreleased]

### Added

* add references to `CITATION.cff`
* function `echo_upper`
* GitHub Action:  automatic CFF validation
* word:  angelo
* word:  formaggio

### Changed

* adjust regex for "studiare" conjugation
* outsource case handling to `echo_upper`
* update `docs-snippets`
* update `libgpl2`

### Fixed

* incorrect CFF file
* typos in docstrings

## [0.2.0] -- 2022-05-01

### Added

* option `--version`

### Changed

* adjust README description
* unite `authors.yaml` with `project.yaml`

### Fixed

* add missing semicolons
* incorrect translations
* typos in changelog items

### Removed

* `authors.yaml`

## [0.1.c] -- 2022-04-28

### Added

* word:  abbiamo
* word:  avere
* word:  avete
* word:  bicicletta
* word:  compito
* word:  conferenza
* word:  conferenze
* word:  corda
* word:  delle
* word:  estate
* word:  forma
* word:  ha
* word:  hai
* word:  hanno
* word:  ho
* word:  informatica
* word:  islandese
* word:  linguistica
* word:  programmatore
* word:  programmatrice
* word:  sala
* word:  sala delle conferenze
* word:  Venezia

### Fixed

* incorrect translations

## [0.1.b] -- 2022-04-28

### Added

* word:  Barcellona
* word:  Berlin
* word:  caffè
* word:  chitarra
* word:  cinema
* word:  dialogo
* word:  gatto
* word:  giornale
* word:  medico
* word:  spaghetti
* word:  zucchero

### Fixed

* too simple translations

## [0.1.a] -- 2022-04-28

### Added

* add citation settings
* submodule `libgpl2`
* word:  ciao
* word:  scusi
* word:  si

### Changed

* link against `libgpl2.a` from corresponding submodule
* update `docs-snippets`
* update `routines`

### Removed

* `GPL_2.0.h`
* `gpl2_copyright_line.c`
* `gpl2_information.c`
* `gpl2_license.c`
* `gpl2_version.c`
* `gpl2_warranty.c`
* `pragma_once.dox`

## [0.1.9] -- 2022-04-26

### Fixed

* add missing semicolon
* incorrect release dates

## [0.1.8] -- 2022-04-26

### Added

* word:  diciannove
* word:  diciassette
* word:  diciotto
* word:  dodici
* word:  dov'
* word:  quattordici
* word:  quindici
* word:  sedici
* word:  settantotto
* word:  tredici
* word:  trentotto
* word:  undici
* word:  venti
* word:  zero

### Fixed

* sort vocabs

## [0.1.7] -- 2022-04-25

### Added

* option `-i`

### Changed

* adjust README description

## [0.1.6] -- 2022-04-23

### Added

* create `gpl2_copyright_line.c`
* create `gpl2_information.c`
* create `gpl2_license.c`
* create `gpl2_version.c`
* create `gpl2_warranty.c`
* create `pragma_once.dox`

### Changed

* move docstrings
* outsource copyright line function to `gpl2_copyright_line.c`
* outsource information function to `gpl2_information.c`
* outsource license function to `gpl2_license.c`
* outsource version function to `gpl2_version.c`
* outsource warranty function to `gpl2_warranty.c`

### Fixed

* missing line breaks
* shorten pattern of "buonasera" / "buona sera"
* shorten pattern of "buongiorno" / "buon giorno"
* sort changelog items

### Removed

* obsolete macros

## [0.1.5] -- 2022-04-23

### Added

* function `license_line`
* global variable `monosyllabic`
* option `-V`

### Changed

* outsource license headline to `license_line`
* set `chatty` to `false` by default

## [0.1.4] -- 2022-04-22

### Added

* function `information`
* function `warranty`
* global variable `chatty`
* macro `GPL_2_0__WARRANTY`
* macro `WARRANTY`
* option `-s`
* option `-w`

### Changed

* adjust README description

## [0.1.3] -- 2022-04-21

### Changed

* adjusted spacing in sections

### Fixed

* improved translation of "piacere"

## [0.1.2] -- 2022-04-21

### Added

* word:  austriaca, austriaco
* word:  buon, buona
* word:  francese
* word:  giorno
* word:  inglese
* word:  polacca, polacco
* word:  sera
* word:  studi
* word:  studia
* word:  studiamo
* word:  studiano
* word:  studiare
* word:  studiate
* word:  studio
* word:  svizzera, svizzero

### Fixed

* duplicate entries for "vola"
* enable alternative spelling for "buonasera"
* enable alternative spelling for "buongiorno"
* mismatching translation for "volano"

## [0.1.1] -- 2022-04-19

### Added

* word:  bello
* word:  chiaro
* word:  davvero
* word:  di
* word:  domanda
* word:  domande
* word:  dove
* word:  interessante
* word:  Londra
* word:  mi
* word:  mirtillo
* word:  no
* word:  non
* word:  perché
* word:  piacere
* word:  sì
* word:  tutto
* word:  vola
* word:  volano
* word:  volare
* word:  volate
* word:  voli
* word:  voliamo
* word:  volo

### Changed

* update `docs-snippets`
* update `routines`

### Fixed

* strings in regex for word "è"
* translation bugs in forms of "chiamarsi"

## [0.1.0] -- 2022-04-14

### Fixed

* missing changelog items
* typos in docstrings
* typos in README description

GPL_2.0.h
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The brief license text of the GPL 2.0 license.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        GPL_2_0.h
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This header defines utility functions and macros concering the GPL 2.0
 * license.
 */

/******************************************************************************/

/*
 * Security settings.
 */

#pragma once
#ifndef __GPL_2_0_H__
#define __GPL_2_0_H__



/*
 * Includes.
 */

// The C standard library.
#include <stdbool.h>
#include <stdio.h>



/**
 * \brief   The GPL 2.0 license (brief version).
 *
 * This macro represents the brief version of the GPL 2.0 license text.
 */

#define GPL_2_0                                                                \
    "This program is free software; you can redistribute it and/or modify\n"   \
    "it under the terms of the GNU General Public License as published by\n"   \
    "the Free Software Foundation; either version 2 of the License, or\n"      \
    "(at your option) any later version.\n"                                    \
    "\n"                                                                       \
    "This program is distributed in the hope that it will be useful,\n"        \
    "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"         \
    "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"          \
    "GNU General Public License for more details.\n"                           \
    "\n"                                                                       \
    "You should have received a copy of the GNU General Public License along\n"\
    "with this program; if not, write to the Free Software Foundation, Inc.,\n"\
    "51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.\n"



/**
 * \brief   The GPL 2.0 warranty.
 *
 * This macro represents the entire warranty section of the GPL 2.0 license.
 */

#define GPL_2_0__WARRANTY                                                       \
"                            NO WARRANTY\n"                                     \
"\n"                                                                            \
"  11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY\n"  \
"FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW.  EXCEPT WHEN\n"    \
"OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES\n"      \
"PROVIDE THE PROGRAM \"AS IS\" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED\n"\
"OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF\n"        \
"MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS\n"   \
"TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU.  SHOULD THE\n"      \
"PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING,\n"    \
"REPAIR OR CORRECTION.\n"                                                       \
"\n"                                                                            \
"  12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING\n" \
"WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR\n"         \
"REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,\n"  \
"INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING\n" \
"OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED\n"   \
"TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY\n"    \
"YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER\n"  \
"PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE\n"       \
"POSSIBILITY OF SUCH DAMAGES.\n"



/*
 * Function declarations.
 */

extern  bool    gpl2_copyright_line ( FILE * const          stream
                                    , const int             year_first
                                    , const int             year_latest
                                    , const char * const    people
                                    );
extern  bool    gpl2_license        ( FILE * const          stream
                                    , const int             year_first
                                    , const int             year_latest
                                    , const char * const    people
                                    );
extern  bool    gpl2_version        ( FILE * const          stream
                                    , const char * const    app_name
                                    , const int             major
                                    , const int             minor
                                    , const int             fix_level
                                    );
extern  bool    gpl2_warranty       (FILE * const stream);



/*
 * End of header.
 */

// Leaving the header.
#endif  // ! __GPL_2_0_H__

/******************************************************************************/


gpl2_license.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 license header.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_license.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 license header function.  It will be written
 * to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Print the copyright line of the GPL 2.0 license header.
 * \param   stream      The stream to write the copyright line to.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * Write the GPL 2.0 license header's copyright line to the given stream.
 */

inline bool gpl2_license ( FILE * const          stream
                         , const int             year_first
                         , const int             year_latest
                         , const char * const    people
                         )
{
    return gpl2_copyright_line (stream, year_first, year_latest, people)
        && fprintf (stream, GPL_2_0);
}

/******************************************************************************/


gpl2_version.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       A minimum version information line.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_version.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements a minimal version information function.  The text will
 * be written to the given stream using **hexadecimal integers**.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   A minimum version information.
 * \param   stream      The stream to write the information to.
 * \param   app_name    The concerning application's name.
 * \param   major       The concerning application's major version.
 * \param   minor       The concerning application's minor version.
 * \param   fix_level   The concerning application's fix level.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a brief version information for the concerning
 * application to the given stream.  Please note that the version will be output
 * using **hexadecimal integer** numbers.
 */

inline bool gpl2_license ( FILE * const          stream
                         , const int             year_first
                         , const int             year_latest
                         , const char * const    people
                         )
{
    return gpl2_copyright_line (stream, year_first, year_latest, people)
        &&             fprintf (stream, GPL_2_0)
         ;
}

/******************************************************************************/


gpl2_warranty.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 warranty information.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_warranty.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 warranty function.  It will write a brief
 * warranty notice to the given stream and than cite the according section of
 * the GPL 2.0 license.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Write the GPL 2.0 warranty information to the given stream.
 * \param   stream  The stream to write the GPL 2.0 warranty information to.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a warranty information to the given stream.
 */

inline bool gpl2_warranty (FILE * const stream)
{
    return fprintf  ( stream
                    , "This application comes with ABSOLUTELY NO WARRANTY, not"
                      " even for MERCHANTABILITY\nor FITNESS FOR A PARTICULAR "
                      "PURPOSE.\n\nThe GPL 2.0 license specifies the following"
                      ":\n\n" GPL_2_0__WARRANTY
                    );
}

/******************************************************************************/


gpl2_information.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 information.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_information.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 information function.  The text will be
 * written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   The information for the interactive mode.
 * \param   stream      The stream to write the information to.
 * \param   app_name    The concerning application's name.
 * \param   major       The concerning application's major version.
 * \param   minor       The concerning application's minor version.
 * \param   fix_level   The concerning application's fix level.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a brief information block to the given stream.  This
 * is useful especially for the interactive mode.
 */

inline bool gpl2_information ( FILE * const          stream
                             , const char * const    app_name
                             , const int             major
                             , const int             minor
                             , const int             fix_level
                             , const int             year_first
                             , const int             year_latest
                             , const char * const    people
                             )
{
    return        gpl2_version ( stream, app_name,   major, minor, fix_level )
        && gpl2_copyright_line ( stream, year_first, year_latest,  people    )
        &&             fprintf ( stream
                               , "This application is free software.  For copyi"
                                 "ng details, see the\nGNU General Public Licen"
                                 "se, version 2, as of June 1991.\n\nThis appli"
                                 "cation comes with ABSOLUTELY NO WARRANTY, not"
                                 " even for MERCHANTABILITY\nor FITNESS FOR A P"
                                 "ARTICULAR PURPOSE; run `%s -w' for more detai"
                                 "ls.\n\n"
                               , app_name
                               )
         ;
}

/******************************************************************************/


gpl2_license_line.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The copyright line of the GPL 2.0 license header.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_license_line.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the copyright line of the GPL 2.0 license header.  It
 * will be written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 *
 */

// .

/******************************************************************************/


gpl2_copyright_line.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The copyright line of the GPL 2.0 license header.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_copyright_line.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the copyright line function of the GPL 2.0 license
 * header.  It will be written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Print the copyright line of the GPL 2.0 license header.
 * \param   stream      The stream to write the copyright line to.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * Write the GPL 2.0 license header's copyright line to the given stream.
 */

inline bool gpl2_copyright_line ( FILE * const          stream
                                , const int             year_first
                                , const int             year_latest
                                , const char * const    people
                                )
{
    return year_first >= year_latest ? fprintf  ( stream
                                                , "Copyright (C) %d %s\n\n"
                                                , year_first
                                                , people
                                                )
                                     : fprintf  ( stream
                                                , "Copyright (C) %d─%d %s\n\n"
                                                , year_first
                                                , year_latest
                                                , people
                                                )
                                     ;
}

/******************************************************************************/


flex-gcc.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      flex-gcc.m
%%
%%  BRIEF
%%      Create a new lexer using Flex and GCC.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
compiler.flags  = [ ' -Wall -Werror -Wextra -Wpedantic '                     ...
                    ' -DYY_NO_INPUT '                                        ...
                    ' -DYY_NO_UNPUT '                                        ...
                  ];
compiler.in     = '*.c';
compiler.link   = [ ' -lfl '                                                 ...
                    ' -L./libgpl2/ -lgpl2 '                                  ...
                  ];
compiler.out    = ['./' 'italiano'];
compiler.self   = 'gcc';
compiler.call   = [ compiler.self ' ' compiler.flags ' ' compiler.in ' '     ...
                    compiler.link ' -o ' compiler.out                        ...
                  ];

octave.dir  = './libgpl2/';
octave.in   = 'gcc-ar.m';
octave.out  = 'libgpl2.a';
octave.self = 'octave';
octave.call = [octave.self ' ' octave.in];

scangen.in      = ['italiano' '.l'];
scangen.out     = 'lex.yy.c';
scangen.self    = 'flex';
scangen.call    = [scangen.self ' ' scangen.in];



% Miscellaneous.
misc.self   = 'flex-gcc.m';
misc.banner = ['[ ' misc.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([misc.banner 'Begin build instruction.']);



% Compile mandatory libraries.
disp ([misc.banner 'Compile mandatory libraries ...']);

directories.self = cd (octave.dir);
system (octave.call);
cd (directories.self);

disp ([misc.banner 'Done.']);



% Clean outdated compilation.
fprintf ([misc.banner 'Remove outdated compilation of ' compiler.out ' ... ']);

if length (glob (compiler.out));
    delete (compiler.out);
end;

disp ('Done.');



% Call Flex.
disp ([misc.banner 'Compile scanner definition with Flex ...']);

disp (scangen.call);
system (scangen.call);

disp ([misc.banner 'Done.']);



% Call GCC.
disp ([misc.banner 'Compile C source code with GCC ...']);

disp (compiler.call);
system (compiler.call);

disp ([misc.banner 'Done.']);



% Clean build artifacts.
fprintf ([misc.banner 'Remove build artifacts ... ']);

if length (glob (scangen.out));
    delete (scangen.out);
end;

disp ('Done.');



% End build instruction.
disp ([misc.banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
